plugins/wallet_plugin/se_wallet.cpp

```diff
@@ -0,0 +1,372 @@
+/**
+ *  @file
+ *  @copyright defined in eos/LICENSE.txt
+ */
+#include <eosio/wallet_plugin/se_wallet.hpp>
+#include <eosio/wallet_plugin/macos_user_auth.h>
+#include <eosio/chain/exceptions.hpp>
+
+#include <fc/crypto/openssl.hpp>
+
+#include <boost/range/adaptor/map.hpp>
+#include <boost/range/algorithm/copy.hpp>
+
+#include <Security/Security.h>
+
+#include <future>
+
+namespace eosio { namespace wallet {
+
+using namespace fc::crypto::r1;
+
+namespace detail {
+
+static void auth_callback(int success, void* data) {
+   promise<bool>* prom = (promise<bool>*)data;
+   prom->set_value(success);
+}
+
+struct se_wallet_impl {
+
+   static public_key_data get_public_key_data(SecKeyRef key) {
+      SecKeyRef pubkey = SecKeyCopyPublicKey(key);
+
+      CFErrorRef error = nullptr;
+      CFDataRef keyrep = nullptr;
+      keyrep = SecKeyCopyExternalRepresentation(pubkey, &error);
+
+      public_key_data pub_key_data;
+      if(!error) {
+         const UInt8* cfdata = CFDataGetBytePtr(keyrep);
+         memcpy(pub_key_data.data+1, cfdata+1, 32);
+         pub_key_data.data[0] = 0x02 + (cfdata[64]&1);
+      }
+
+      CFRelease(keyrep);
+      CFRelease(pubkey);
+
+      if(error) {
+         string error_string = string_for_cferror(error);
+         CFRelease(error);
+         FC_THROW_EXCEPTION(chain::wallet_exception, "Failed to get public key from Secure Enclave: ${m}", ("m", error_string));
+      }
+
+      return pub_key_data;
+   }
+
+   static public_key_type get_public_key(SecKeyRef key) {
```

Comment:
This is somewhat icky, but public_key class doesn’t have a way to conjure up an instance otherwise (unless going through private_key, which of course we don’t have in this case!) There is something similar for a signature.

An alternative would be for us to make a public_key & signature_type ctor that allows me to create said objects with raw x/y and r/s values. Thoughts?

Reply:
Providing an way to do this if you know the wrapped type (like R1) makes sense.  I'm not sure given that you are dealing with a wrapper and stubs in `public_key_type` that it makes sense to suggest there should always be a way to take raw values and form a type.